tools/esp8266/hmRadio.h

```diff
@@ -159,6 +159,39 @@ class HmRadio {
             return mRfChLst[mTxChIdx];
         }
 
+        void sendControlPacket(uint64_t invId, uint8_t cmd, uint16_t data) {
+            DPRINTLN(DBG_VERBOSE, F("hmRadio.h:sendControlPacket"));
+            // sendCmdPacket(invId, 0x51, 0x80, false); // 0x80 implementation as original DTU code
+            sendCmdPacket(invId, 0x51, 0x81, false);
+            int cnt = 0;
+            mTxBuf[10] = cmd; // cmd --> 0x0b => Type_ActivePowerContr, 0 on, 1 off, 2 restart, 12 reactive power, 13 power factor
+            mTxBuf[10 + (++cnt)] = 0x00;
+            if (cmd == 11){
+                // 4 bytes control data
+                // Power Limit fix point 10 eg. 30 W --> 0d300 = 0x012c
+                // -1 = 0xffff --> no limit
+                if (data == 0xffff){
+                    data &= 0xffff; // ToDo: unlimit value is needed and is inverter specific! --> get it via RF from inverter or via user interface
+                } else {
+                    data*= 10;
+                }
+                mTxBuf[10 + (++cnt)] = (data >> 8) & 0xff; // 0x01
+                mTxBuf[10 + (++cnt)] = (data     ) & 0xff; // 0x2c
+                // mTxBuf[10 + (++cnt)] = 0x00; // not persistent
+                mTxBuf[10 + (++cnt)] = 0x01; // persistent
```

Comment:
Sollte m.E. 0x00 0x01 heißen, laut Original Source Code.
Laut unseren bisherigen Traces ist der Default 0x00 0x00.

Reply:
@stefan123t 
> 1, ich wäre für zwei/drei enums in denen die ganzen MainCmd's und die SubCmd's für DevControl und DevInformation aufgeführt sind. Dann kann man die Namen benutzen und muß nicht immer mit den Hex Werten hantieren.
--> Jawohl das ist eine gute Idee. Können wir ja zusammen nun im branch 'pwrlimit' machen.

> Zu 2.: Ja das mit den MQTT topics ist nicht intuitiv. Die Intension war, dass man dadurch sehr effizient die weiteren Kommandos implementieren/testen kann und stupide einfach die Kommandos als topic übergibt. Manchmal mit, manchmal ohne Payload. Mein Gedanke war auch dass man alles nur an ein topic pro inverter hängt und eine json payload baut. Diese könnte man dann auch 1:1 in einem Web-Inteface verwenden.
Die Abstimmung mit dem Modbus interface ist natürlich sinnvoll. Wobei ich hier zunächst 1:1 die Implementierung der hoymiles DTU pro Gen3 clonen würde. Ich hatte da einfach mal eine Email zum Hersteller geschrieben und die haben mir ein PDF mit den Details gesendet. (Ich denke aber das Dokument ist auch öffentlich im I-Net bekannt.)
--> Also Vorschlag:
- MQTT und Web Interface zu Steuerung sollten die gleiche Logik haben. Man muss auch die Anknüpfung an OpenHab, Homeassistent beachten damit man hier nicht unnötig viel tun muss. (Aus dieser Sicht ist es am Ende egal ob 0 Ein bedeutet oder 1 Aus. Das ist dann alles in einer GUI / APP)
- Das Cmd "11, ActivePowerLimit" bekommt neben der eigentlichen Payload in Watt noch eine zweite Payload 'persistent' wird per MQTT/Web Interface über geben und in der oben zitierten Funktion als "data2" übergeben. Einfach so z.B. "400|1" für remanent und "400|0" für temporär. Der Aufrauf von sendControlPacket(uint64_t invId, uint8_t cmd, uint16_t data)  wird dann eben zu sendControlPacket(uint64_t invId, uint8_t cmd, uint16_t data, **uint16_data2**)  Dann muss man eben für dieses ( und auch für die Anderen wie ReactivePower) ein mapping bauen im User Frontend.
- Modbus Interface wird identisch zu DTU pro implementiert.

Zu 4. PowerPFDev.Desc in hmRadio.h sollte eigentlich default 0x00 00 sein. --> Ja, default Null ist besser wenn nichts anderes gewünscht plus die Option, dass man per Interface senden kann bspw. "400 Watt Limit remanent" (s.o.). Noch was anderes Ich denke auch das der Begriff/Wort "PowerPFDev" --> "Sollwert für den Leistungsfaktor" in diesem Kontext falsch ist. Ich denke wenn man die verfügbaren Code genau durchforstet ist dass der Wert für die Payload / Data den man dem SubCmd 13 übergibt. Oder? 

Zu 5.
> * mClient wird public, das hatte mich auch schon gewundert.
--> Ja das war um abzukürzen. Ist nicht gut. Die Anforderung ist aber dass man in main eine Subscrition auf ein Topic auslösen kann.
Vorschlag: 
Implementieren einer entsprechenden Funktion in der mqtt Klasse damit der mClient wieder private wird und man über die Public Funktion eine Subscribtion machen kann.

Alle Vorschlägen meinerseits können / sollten im branch 'pwrlimit' eingebaut / verbessert werden und dann am Ende des Sprints ein Review und die Entscheidung "Release in main" oder "weiterer Sprint"